File: javascript/tweet.js
require('dotenv').config();
const { chromium } = require('playwright-extra');
const stealth = require('puppeteer-extra-plugin-stealth')();
const path = require('path');
const fs = require('fs');

// Add stealth plugin
chromium.use(stealth);

const AUTH_STATE_FILE = path.join(__dirname, 'twitter-auth.json');
const COOKIES_FILE = path.join(__dirname, 'twitter_cookies.json');

// Convert browser extension cookies to Playwright format
function convertCookies(rawCookies) {
    return rawCookies.map(cookie => {
        let sameSite = cookie.sameSite || 'Lax';
        
        // Normalize sameSite values
        if (!['Strict', 'Lax', 'None'].includes(sameSite)) {
            if (sameSite === 'no_restriction') {
                sameSite = 'None';
            } else if (sameSite === 'lax') {
                sameSite = 'Lax';
            } else if (sameSite === 'strict') {
                sameSite = 'Strict';
            } else {
                sameSite = 'Lax';
            }
        }

        return {
            name: cookie.name,
            value: cookie.value,
            domain: cookie.domain,
            path: cookie.path || '/',
            expires: cookie.expirationDate || -1,
            httpOnly: cookie.httpOnly || false,
            secure: cookie.secure || false,
            sameSite: sameSite
        };
    });
}

async function loginManually() {
    console.log('Opening browser for manual login...');
    const browser = await chromium.launch({ 
        headless: false,
        args: [
            '--disable-blink-features=AutomationControlled',
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage'
        ]
    });
    
    const context = await browser.newContext({
        viewport: { width: 1280, height: 720 },
        userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
        locale: 'en-US',
        timezoneId: 'America/New_York'
    });
    
    const page = await context.newPage();
    
    await page.goto('https://twitter.com/login');
    
    console.log('\n=================================================');
    console.log('PLEASE LOG IN MANUALLY IN THE BROWSER WINDOW');
    console.log('After successful login, close the browser or press Ctrl+C');
    console.log('Waiting 120 seconds for you to complete login...');
    console.log('=================================================\n');
    
    // Wait for user to login manually
    await page.waitForTimeout(120000);
    
    // Save the authentication state
    await context.storageState({ path: AUTH_STATE_FILE });
    console.log(`\n✓ Session saved to ${AUTH_STATE_FILE}`);
    console.log('You can now use the saved session to send tweets!\n');

    await context.close();
    await browser.close();
}

async function sendTweetWithCookies(message) {
    if (!fs.existsSync(COOKIES_FILE)) {
        console.error(`Error: ${COOKIES_FILE} not found!`);
        console.log('\nHow to export cookies:');
        console.log('1. Install "Cookie-Editor" extension in your browser');
        console.log('2. Go to twitter.com and make sure you\'re logged in');
        console.log('3. Click the Cookie-Editor extension icon');
        console.log('4. Click "Export" -> "Export as JSON"');
        console.log(`5. Save the file as "${COOKIES_FILE}" in this directory`);
        console.log('\nAlternatively, you can use "EditThisCookie" or similar extensions.');
        return;
    }

    console.log(`Loading cookies from ${COOKIES_FILE}...`);
    const rawCookies = JSON.parse(fs.readFileSync(COOKIES_FILE, 'utf8'));
    const cookies = convertCookies(rawCookies);

    const browser = await chromium.launch({ 
        headless: false,
        args: [
            '--disable-blink-features=AutomationControlled',
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage'
        ]
    });
    
    const context = await browser.newContext({
        viewport: { width: 1280, height: 720 },
        userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
        locale: 'en-US',
        timezoneId: 'America/New_York'
    });
    
    // Add cookies to context
    await context.addCookies(cookies);
    
    const page = await context.newPage();

    console.log('Navigating to Twitter...');
    await page.goto('https://twitter.com/home');
    await page.waitForTimeout(3000);
    
    // Close any popups if they appear
    try {
        await page.click('[data-testid="app-bar-close"]', { timeout: 2000 });
    } catch (e) {
        // Popup not present
    }
    
    console.log('Composing tweet...');
    await page.fill('[data-testid="tweetTextarea_0"]', message);
    await page.waitForTimeout(1000);
    
    console.log('Posting tweet...');
    await page.click('[data-testid="tweetButtonInline"]');
    
    console.log('✓ Tweet posted successfully!');
    await page.waitForTimeout(3000);

    await context.close();
    await browser.close();
}

async function sendTweet(message) {
    if (!fs.existsSync(AUTH_STATE_FILE)) {
        console.error(`Error: ${AUTH_STATE_FILE} not found!`);
        console.log('Please run "node tweet.js login" first to save your session.');
        return;
    }

    console.log('Loading saved session...');
    const browser = await chromium.launch({ 
        headless: false,
        args: [
            '--disable-blink-features=AutomationControlled',
            '--no-sandbox',
            '--disable-setuid-sandbox',
            '--disable-dev-shm-usage'
        ]
    });
    
    const context = await browser.newContext({
        storageState: AUTH_STATE_FILE,
        viewport: { width: 1280, height: 720 },
        userAgent: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36',
        locale: 'en-US',
        timezoneId: 'America/New_York'
    });

    const page = await context.newPage();

    console.log('Navigating to tweet composer...');
    await page.goto('https://twitter.com/compose/tweet', { waitUntil: 'networkidle' });
    await page.waitForTimeout(3000);
    
    console.log('Typing tweet message...');
    const textarea = await page.locator('[data-testid="tweetTextarea_0"]');
    await textarea.click();
    await page.waitForTimeout(500);
    await textarea.type(message, { delay: 100 });
    await page.waitForTimeout(1000);

    console.log('Clicking tweet button...');
    await page.click('[data-testid="tweetButtonInline"]');
    
    console.log('Tweet posted! Waiting to confirm...');
    await page.waitForTimeout(3000);
    
    await browser.close();
    console.log('Done!');
}

// Main execution
const command = process.argv[2];
const message = process.argv[3];

if (!command) {
    console.log('Usage:');
    console.log('  Login:           node tweet.js login');
    console.log('  Post (saved):    node tweet.js post "Your message"');
    console.log('  Post (cookies):  node tweet.js cookies "Your message"');
    process.exit(1);
}

(async () => {
    try {
        if (command === 'login') {
            await loginManually();
        } else if (command === 'post') {
            if (!message) {
                console.error('Error: Please provide a message to post');
                console.log('Usage: node tweet.js post "Your message here"');
                process.exit(1);
            }
            await sendTweet(message);
        } else if (command === 'cookies') {
            if (!message) {
                console.error('Error: Please provide a message to post');
                console.log('Usage: node tweet.js cookies "Your message here"');
                process.exit(1);
            }
            await sendTweetWithCookies(message);
        } else {
            console.error(`Unknown command: ${command}`);
            console.log('Available commands: login, post, cookies');
            process.exit(1);
        }
    } catch (err) {
        console.error('Error:', err.message);
        process.exit(1);
    }
})();

File: python/tweet.py
import os
import sys
import json
from playwright.sync_api import Playwright, sync_playwright
from dotenv import load_dotenv

load_dotenv()

USERNAME = "mrcentimetre@gmail.com"
PASSWORD = "Centi@2001#"
AUTH_FILE = "twitter.json"
COOKIES_FILE = "twitter_cookies.json"


def login(playwright: Playwright) -> None:
    """Login to Twitter and save session"""
    print("Opening browser for manual login...")
    browser = playwright.chromium.launch(headless=False)
    context = browser.new_context(
        viewport={'width': 1280, 'height': 720},
        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36'
    )
    page = context.new_page()

    page.goto("https://twitter.com/i/flow/login")
    
    print("\n=================================================")
    print("PLEASE LOG IN MANUALLY IN THE BROWSER WINDOW")
    print("After successful login, close the browser window")
    print("Or press Ctrl+C when done")
    print("=================================================\n")
    
    # Wait for user to login manually
    try:
        page.wait_for_timeout(120000)  # Wait 2 minutes
    except:
        pass
    
    # Save the session
    context.storage_state(path=AUTH_FILE)
    print(f"\n✓ Session saved to {AUTH_FILE}")
    print("You can now use the 'post' command to send tweets automatically!\n")

    context.close()
    browser.close()


def post(playwright: Playwright, message: str) -> None:
    """Post a tweet using saved session"""
    if not os.path.exists(AUTH_FILE):
        print(f"Error: {AUTH_FILE} not found. Please run login first:")
        print("  python tweet_python.py login")
        return

    print("Loading saved session...")
    browser = playwright.chromium.launch(headless=False)
    context = browser.new_context(storage_state=AUTH_FILE)
    page = context.new_page()

    print("Navigating to Twitter...")
    page.goto("https://twitter.com/home")
    page.wait_for_timeout(2000)
    
    # Close any popups if they appear
    try:
        page.get_by_test_id("app-bar-close").click(timeout=2000)
    except:
        pass
    
    print("Composing tweet...")
    page.get_by_test_id("tweetTextarea_0").fill(message)
    page.wait_for_timeout(1000)
    
    print("Posting tweet...")
    page.get_by_test_id("tweetButtonInline").click()
    
    print("✓ Tweet posted successfully!")
    page.wait_for_timeout(3000)

    context.close()
    browser.close()


def post_with_browser(playwright: Playwright, message: str) -> None:
    """Post a tweet using your existing Chrome browser profile"""
    # Chrome user data directory on macOS
    user_data_dir = os.path.expanduser("~/Library/Application Support/Google/Chrome")
    
    if not os.path.exists(user_data_dir):
        print("Error: Chrome profile not found. Make sure Chrome is installed.")
        print("Trying alternative path...")
        user_data_dir = os.path.expanduser("~/Library/Application Support/Chromium")
        if not os.path.exists(user_data_dir):
            print("No browser profile found. Use 'login' command instead.")
            return
    
    print("Opening your Chrome browser with existing cookies...")
    context = playwright.chromium.launch_persistent_context(
        user_data_dir,
        headless=False,
        channel="chrome",  # Use installed Chrome instead of Chromium
        args=['--disable-blink-features=AutomationControlled']
    )
    
    page = context.new_page()

    print("Navigating to Twitter...")
    page.goto("https://twitter.com/home")
    page.wait_for_timeout(3000)
    
    # Close any popups if they appear
    try:
        page.get_by_test_id("app-bar-close").click(timeout=2000)
    except:
        pass
    
    print("Composing tweet...")
    page.get_by_test_id("tweetTextarea_0").fill(message)
    page.wait_for_timeout(1000)
    
    print("Posting tweet...")
    page.get_by_test_id("tweetButtonInline").click()
    
    print("✓ Tweet posted successfully!")
    page.wait_for_timeout(3000)

    context.close()


def post_with_cookies(playwright: Playwright, message: str) -> None:
    """Post a tweet using cookies exported from browser"""
    if not os.path.exists(COOKIES_FILE):
        print(f"Error: {COOKIES_FILE} not found!")
        print("\nHow to export cookies:")
        print("1. Install 'Cookie-Editor' extension in your browser")
        print("2. Go to twitter.com and make sure you're logged in")
        print("3. Click the Cookie-Editor extension icon")
        print("4. Click 'Export' -> 'Export as JSON'")
        print(f"5. Save the file as '{COOKIES_FILE}' in this directory")
        print("\nAlternatively, you can use 'EditThisCookie' or similar extensions.")
        return

    print(f"Loading cookies from {COOKIES_FILE}...")
    with open(COOKIES_FILE, 'r') as f:
        cookies_raw = json.load(f)
    
    # Convert cookies to Playwright format
    cookies = []
    for cookie in cookies_raw:
        # Handle sameSite attribute
        same_site = cookie.get('sameSite', 'Lax')
        if same_site not in ['Strict', 'Lax', 'None']:
            if same_site == 'no_restriction':
                same_site = 'None'
            elif same_site == 'lax':
                same_site = 'Lax'
            elif same_site == 'strict':
                same_site = 'Strict'
            else:
                same_site = 'Lax'
        
        playwright_cookie = {
            'name': cookie.get('name'),
            'value': cookie.get('value'),
            'domain': cookie.get('domain'),
            'path': cookie.get('path', '/'),
            'expires': cookie.get('expirationDate', -1),
            'httpOnly': cookie.get('httpOnly', False),
            'secure': cookie.get('secure', False),
            'sameSite': same_site
        }
        cookies.append(playwright_cookie)
    
    browser = playwright.chromium.launch(headless=False)
    context = browser.new_context(
        viewport={'width': 1280, 'height': 720},
        user_agent='Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36'
    )
    
    # Add cookies to context
    context.add_cookies(cookies)
    
    page = context.new_page()

    print("Navigating to Twitter...")
    page.goto("https://twitter.com/home")
    page.wait_for_timeout(3000)
    
    # Close any popups if they appear
    try:
        page.get_by_test_id("app-bar-close").click(timeout=2000)
    except:
        pass
    
    print("Composing tweet...")
    page.get_by_test_id("tweetTextarea_0").fill(message)
    page.wait_for_timeout(1000)
    
    print("Posting tweet...")
    page.get_by_test_id("tweetButtonInline").click()
    
    print("✓ Tweet posted successfully!")
    page.wait_for_timeout(3000)

    context.close()
    browser.close()


if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage:")
        print("  Login:           python3 tweet_python.py login")
        print("  Post (saved):    python3 tweet_python.py post 'Your message'")
        print("  Post (cookies):  python3 tweet_python.py cookies 'Your message'")
        print("  Post (browser):  python3 tweet_python.py browser 'Your message'")
        sys.exit(1)
    
    command = sys.argv[1]
    
    with sync_playwright() as playwright:
        if command == "login":
            login(playwright)
        elif command == "post":
            if len(sys.argv) < 3:
                print("Error: Please provide a message to post")
                print("Usage: python3 tweet_python.py post 'Your message here'")
                sys.exit(1)
            message = sys.argv[2]
            post(playwright, message)
        elif command == "cookies":
            if len(sys.argv) < 3:
                print("Error: Please provide a message to post")
                print("Usage: python3 tweet_python.py cookies 'Your message here'")
                sys.exit(1)
            message = sys.argv[2]
            post_with_cookies(playwright, message)
        elif command == "browser":
            if len(sys.argv) < 3:
                print("Error: Please provide a message to post")
                print("Usage: python3 tweet_python.py browser 'Your message here'")
                sys.exit(1)
            message = sys.argv[2]
            post_with_browser(playwright, message)
        else:
            print(f"Unknown command: {command}")
            print("Available commands: login, post, cookies, browser")
            sys.exit(1)
